File: src/treeline/infra/csv_provider.py
"""CSV file provider for importing transactions."""

import csv
import re
from datetime import date, datetime, timezone
from decimal import Decimal, InvalidOperation
from pathlib import Path
from typing import Any, Dict, List, Tuple
from uuid import uuid4

from treeline.abstractions import DataAggregationProvider
from treeline.domain import Account, BalanceSnapshot, Fail, Ok, Result, Transaction


class CSVProvider(DataAggregationProvider):
    """CSV file implementation for data aggregation."""

    @property
    def can_get_accounts(self) -> bool:
        return False  # Accounts must be manually mapped

    @property
    def can_get_transactions(self) -> bool:
        return True

    @property
    def can_get_balances(self) -> bool:
        return False

    async def get_accounts(
        self,
        provider_account_ids: List[str] = [],
        provider_settings: Dict[str, Any] | None = None,
    ) -> Result[List[Account]]:
        """CSV provider does not support getting accounts."""
        return Fail("CSV provider does not support getting accounts")

    async def get_transactions(
        self,
        start_date: datetime,
        end_date: datetime,
        provider_account_ids: List[str] = [],
        provider_settings: Dict[str, Any] | None = None,
    ) -> Result[List[Transaction]]:
        """Parse CSV file and return transactions."""
        if not provider_settings:
            return Fail("provider_settings is required")

        file_path = provider_settings.get("file_path")
        if not file_path:
            return Fail("file_path is required in provider_settings")

        column_mapping = provider_settings.get("column_mapping")
        if not column_mapping:
            return Fail("column_mapping is required in provider_settings")

        date_format = provider_settings.get("date_format", "auto")
        flip_signs = provider_settings.get("flip_signs", False)
        debit_negative = provider_settings.get("debit_negative", False)

        # Check if file exists
        path = Path(file_path)
        if not path.exists():
            return Fail(f"File not found: {file_path}")

        try:
            with open(path, "r", encoding="utf-8") as f:
                reader = csv.DictReader(f)
                transactions = []

                for row in reader:
                    # Parse transaction from CSV row
                    tx_result = self._parse_transaction_row(
                        row, column_mapping, date_format, debit_negative
                    )
                    if not tx_result.success:
                        # Skip invalid rows but continue processing
                        continue

                    tx = tx_result.data

                    # Apply sign flip if requested
                    if flip_signs:
                        tx = tx.model_copy(update={"amount": -tx.amount})

                    transactions.append(tx)

                return Ok(transactions)

        except Exception as e:
            return Fail(f"Failed to parse CSV file: {str(e)}")

    async def get_balances(
        self,
        provider_account_ids: List[str] = [],
        provider_settings: Dict[str, Any] | None = None,
    ) -> Result[List[BalanceSnapshot]]:
        """CSV provider does not support getting balances."""
        return Fail("CSV provider does not support getting balances")

    def _parse_transaction_row(
        self,
        row: Dict[str, str],
        column_mapping: Dict[str, str],
        date_format: str,
        debit_negative: bool = False,
    ) -> Result[Transaction]:
        """Parse a single CSV row into a Transaction."""
        try:
            # Get column names from mapping
            date_col = column_mapping.get("date")
            description_col = column_mapping.get("description")
            amount_col = column_mapping.get("amount")
            debit_col = column_mapping.get("debit")
            credit_col = column_mapping.get("credit")
            posted_date_col = column_mapping.get("posted_date")

            if not date_col:
                return Fail("date column is required in column_mapping")

            # Must have either amount OR (debit and credit)
            has_amount = bool(amount_col)
            has_debit_credit = bool(debit_col or credit_col)

            if not has_amount and not has_debit_credit:
                return Fail(
                    "amount column (or debit/credit columns) required in column_mapping"
                )

            # Parse date
            date_str = row.get(date_col, "").strip()
            if not date_str:
                return Fail("Missing date value")

            transaction_date = self._parse_date(date_str, date_format)
            if not transaction_date:
                return Fail(f"Failed to parse date: {date_str}")

            # Parse posted_date if provided
            if posted_date_col:
                posted_date_str = row.get(posted_date_col, "").strip()
                if posted_date_str:
                    posted_date = self._parse_date(posted_date_str, date_format)
                    if not posted_date:
                        posted_date = transaction_date
                else:
                    posted_date = transaction_date
            else:
                posted_date = transaction_date

            # Parse amount - handle both single amount and debit/credit columns
            if has_amount:
                amount_str = row.get(amount_col, "").strip()
                if not amount_str:
                    return Fail("Missing amount value")

                amount = self._parse_amount(amount_str)
                if amount is None:
                    return Fail(f"Failed to parse amount: {amount_str}")
            else:
                # Handle debit/credit columns
                # Use whichever column has a value (they're mutually exclusive in most CSVs)
                debit_str = row.get(debit_col, "").strip() if debit_col else ""
                credit_str = row.get(credit_col, "").strip() if credit_col else ""

                # Skip row if both are empty
                if not debit_str and not credit_str:
                    return Fail("Both debit and credit are empty")

                # Parse values
                debit_amt = self._parse_amount(debit_str) if debit_str else None
                credit_amt = self._parse_amount(credit_str) if credit_str else None

                if debit_amt is not None and credit_amt is not None:
                    # Both have values - this is unusual but handle it
                    # Take the non-zero one, or use the larger absolute value
                    if abs(debit_amt) > abs(credit_amt):
                        amount = debit_amt  # Preserve sign from CSV
                    else:
                        amount = credit_amt  # Preserve sign from CSV
                elif debit_amt is not None:
                    # Only debit has value - preserve sign from CSV, then apply debit_negative
                    amount = debit_amt
                    if debit_negative and amount > 0:
                        amount = -amount
                elif credit_amt is not None:
                    # Only credit has value - preserve sign from CSV
                    amount = credit_amt
                else:
                    return Fail(
                        f"Failed to parse debit/credit: {debit_str}/{credit_str}"
                    )

            # Parse description and clean it
            description = ""
            if description_col:
                raw_description = row.get(description_col, "").strip()
                description = self._clean_description(raw_description)

            # Create transaction
            # Note: account_id will be set by ImportService when mapping to target account
            transaction = Transaction(
                id=uuid4(),
                account_id=uuid4(),  # Placeholder, will be replaced by ImportService
                amount=amount,
                description=description,
                transaction_date=transaction_date,
                posted_date=posted_date,
                created_at=datetime.now(timezone.utc),
                updated_at=datetime.now(timezone.utc),
            )

            return Ok(transaction)

        except Exception as e:
            return Fail(f"Failed to parse transaction row: {str(e)}")

    def _parse_date(self, date_str: str, date_format: str) -> date | None:
        """Parse date string and return date object (no timezone)."""
        if not date_str:
            return None

        # Try auto-detection if format is "auto"
        if date_format == "auto":
            formats = [
                "%Y-%m-%d",  # 2024-10-01
                "%m/%d/%Y",  # 10/01/2024
                "%d/%m/%Y",  # 01/10/2024
                "%Y/%m/%d",  # 2024/10/01
                "%m-%d-%Y",  # 10-01-2024
                "%d-%m-%Y",  # 01-10-2024
            ]
        else:
            # Map common format names to strftime formats
            format_map = {
                "YYYY-MM-DD": "%Y-%m-%d",
                "MM/DD/YYYY": "%m/%d/%Y",
                "DD/MM/YYYY": "%d/%m/%Y",
                "YYYY/MM/DD": "%Y/%m/%d",
            }
            fmt = format_map.get(date_format)
            if not fmt:
                return None
            formats = [fmt]

        # Try each format
        for fmt in formats:
            try:
                dt = datetime.strptime(date_str, fmt)
                # Return date object, not datetime (no timezone conversion)
                return dt.date()
            except ValueError:
                continue

        return None

    def _parse_amount(self, amount_str: str) -> Decimal | None:
        """Parse amount string, handling $ signs and commas."""
        if not amount_str:
            return None

        try:
            # Remove currency symbols and commas
            cleaned = amount_str.strip()
            cleaned = cleaned.replace("$", "")
            cleaned = cleaned.replace(",", "")
            cleaned = cleaned.replace(" ", "")

            # Handle parentheses notation for negative numbers: (100.00) -> -100.00
            if cleaned.startswith("(") and cleaned.endswith(")"):
                cleaned = "-" + cleaned[1:-1]

            return Decimal(cleaned)
        except (InvalidOperation, ValueError):
            return None

    def _clean_description(self, description: str) -> str:
        """Remove CSV noise from descriptions before storing.

        Removes:
        - Literal "null" strings
        - Card number masks (XXXXXXXXXXXX1234)
        - Extra whitespace
        """
        if not description:
            return ""

        cleaned = description

        # Remove literal "null" strings (case insensitive)
        cleaned = re.sub(r"\bnull\b", "", cleaned, flags=re.IGNORECASE)

        # Remove card number masks (XXXXXXXXXXXX followed by digits)
        cleaned = re.sub(r"x{10,}\d+", "", cleaned, flags=re.IGNORECASE)

        # Clean up extra whitespace (collapse multiple spaces, trim)
        cleaned = re.sub(r"\s+", " ", cleaned).strip()

        return cleaned

    def detect_columns(self, file_path: str) -> Result[Dict[str, str]]:
        """Auto-detect column mapping from CSV headers.

        Returns best-guess mapping for date, amount, and description columns.
        """
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                reader = csv.DictReader(f)
                headers = reader.fieldnames or []

            # Expanded patterns for column detection with fuzzy matching
            date_patterns = [
                "date",
                "transaction date",
                "trans date",
                "txn date",
                "txndate",
                "posted",
                "post date",
                "dt",  # Abbreviated
            ]
            desc_patterns = [
                "description",
                "desc",  # Abbreviated
                "memo",
                "payee",
                "merchant",
                "details",
                "narration",  # International
            ]
            amount_patterns = [
                "amount",
                "amt",  # Abbreviated
                "total",
                "transaction amount",
            ]
            debit_patterns = [
                "debit",
                "dr",  # International abbreviation
                "withdrawal",
                "debit amount",
            ]
            credit_patterns = [
                "credit",
                "cr",  # International abbreviation
                "deposit",
                "credit amount",
            ]

            detected = {}

            # Find date column
            for header in headers:
                header_lower = header.lower().strip()
                if any(pattern in header_lower for pattern in date_patterns):
                    detected["date"] = header
                    break

            # Find amount column (prefer single amount column) - with fuzzy matching
            for header in headers:
                header_lower = header.lower().strip()
                # Remove common suffixes like currency codes for fuzzy matching
                header_clean = re.sub(r"\s+(usd|eur|gbp|cad|aud)$", "", header_lower)
                if any(pattern in header_clean for pattern in amount_patterns):
                    detected["amount"] = header
                    break

            # If no 'amount' found, check for debit/credit
            if "amount" not in detected:
                debit_col = None
                credit_col = None
                for header in headers:
                    header_lower = header.lower().strip()
                    if any(pattern in header_lower for pattern in debit_patterns):
                        debit_col = header
                    if any(pattern in header_lower for pattern in credit_patterns):
                        credit_col = header

                if debit_col or credit_col:
                    detected["debit"] = debit_col
                    detected["credit"] = credit_col

            # Find description column with fallback
            for header in headers:
                header_lower = header.lower().strip()
                # Skip if this is a date column
                if detected.get("date") and header == detected["date"]:
                    continue
                if any(pattern in header_lower for pattern in desc_patterns):
                    detected["description"] = header
                    break

            # If no description found, try fallback columns
            if "description" not in detected:
                fallback_patterns = ["name", "type", "ref", "reference", "category"]
                for header in headers:
                    header_lower = header.lower().strip()
                    # Skip if this is a date column
                    if detected.get("date") and header == detected["date"]:
                        continue
                    if any(pattern in header_lower for pattern in fallback_patterns):
                        detected["description"] = header
                        break

            return Ok(detected)

        except Exception as e:
            return Fail(f"Failed to detect columns: {str(e)}")

    def should_negate_debits(
        self, file_path: str, debit_col: str, credit_col: str
    ) -> Result[bool]:
        """Detect if debit values should be negated (unsigned debit/credit convention).

        Returns True if debits appear to be unsigned positive values that should be negated.
        Returns False if debits are already signed (negative) or mixed.
        """
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                reader = csv.DictReader(f)

                debit_values = []
                for i, row in enumerate(reader):
                    if i >= 10:  # Sample first 10 rows
                        break

                    debit_str = row.get(debit_col, "").strip()
                    if debit_str:
                        debit_amt = self._parse_amount(debit_str)
                        if debit_amt is not None:
                            debit_values.append(debit_amt)

                # If we have debit values and they're all positive, suggest negating
                if len(debit_values) >= 2:  # Need at least 2 samples
                    all_positive = all(amt > 0 for amt in debit_values)
                    return Ok(all_positive)

                # Not enough data to determine
                return Ok(False)

        except Exception as e:
            return Fail(f"Failed to analyze debit convention: {str(e)}")

    def preview_transactions(
        self,
        file_path: str,
        column_mapping: Dict[str, str],
        date_format: str = "auto",
        limit: int = 5,
        flip_signs: bool = False,
        debit_negative: bool = False,
    ) -> Result[List[Transaction]]:
        """Preview first N transactions from CSV with given mapping.

        This is used to show the user what will be imported before committing.
        """
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                reader = csv.DictReader(f)
                transactions = []

                for i, row in enumerate(reader):
                    if i >= limit:
                        break

                    # Parse transaction
                    tx_result = self._parse_transaction_row(
                        row, column_mapping, date_format, debit_negative
                    )

                    if tx_result.success and tx_result.data:
                        tx = tx_result.data

                        # Apply sign flip if requested
                        if flip_signs:
                            tx = tx.model_copy(update={"amount": -tx.amount})

                        transactions.append(tx)

                return Ok(transactions)

        except Exception as e:
            return Fail(f"Failed to preview transactions: {str(e)}")

File: cli-rs/src/infra/csv_provider.rs
//! CSV provider for importing transactions from files.

use crate::domain::{ServiceResult, Transaction};
use chrono::{NaiveDate, Utc};
use regex::Regex;
use rust_decimal::Decimal;
use std::collections::HashMap;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;
use std::str::FromStr;
use uuid::Uuid;

#[derive(Debug, Clone)]
pub struct ColumnMapping {
    pub date: Option<String>,
    pub description: Option<String>,
    pub amount: Option<String>,
    pub debit: Option<String>,
    pub credit: Option<String>,
    pub posted_date: Option<String>,
}

impl ColumnMapping {
    pub fn new() -> Self {
        ColumnMapping {
            date: None,
            description: None,
            amount: None,
            debit: None,
            credit: None,
            posted_date: None,
        }
    }
}

pub struct CSVProvider;

impl CSVProvider {
    pub fn new() -> Self {
        CSVProvider
    }

    /// Detect column mapping from CSV headers
    pub fn detect_columns(file_path: &str) -> ServiceResult<ColumnMapping> {
        let path = Path::new(file_path);
        if !path.exists() {
            return ServiceResult::fail(format!("File not found: {}", file_path));
        }

        let file = match File::open(path) {
            Ok(f) => f,
            Err(e) => return ServiceResult::fail(format!("Failed to open file: {}", e)),
        };

        let mut reader = csv::Reader::from_reader(BufReader::new(file));
        let headers: Vec<String> = match reader.headers() {
            Ok(h) => h.iter().map(|s| s.to_string()).collect(),
            Err(e) => return ServiceResult::fail(format!("Failed to read headers: {}", e)),
        };

        let date_patterns = ["date", "transaction date", "trans date", "txn date", "posted", "post date"];
        let desc_patterns = ["description", "desc", "memo", "payee", "merchant", "details", "narration"];
        let amount_patterns = ["amount", "amt", "total", "transaction amount"];
        let debit_patterns = ["debit", "dr", "withdrawal", "debit amount"];
        let credit_patterns = ["credit", "cr", "deposit", "credit amount"];

        let mut mapping = ColumnMapping::new();

        // Find date column
        for header in &headers {
            let lower = header.to_lowercase();
            if date_patterns.iter().any(|p| lower.contains(p)) {
                mapping.date = Some(header.clone());
                break;
            }
        }

        // Find amount column
        for header in &headers {
            let lower = header.to_lowercase();
            if amount_patterns.iter().any(|p| lower.contains(p)) {
                mapping.amount = Some(header.clone());
                break;
            }
        }

        // If no amount, look for debit/credit
        if mapping.amount.is_none() {
            for header in &headers {
                let lower = header.to_lowercase();
                if debit_patterns.iter().any(|p| lower.contains(p)) {
                    mapping.debit = Some(header.clone());
                }
                if credit_patterns.iter().any(|p| lower.contains(p)) {
                    mapping.credit = Some(header.clone());
                }
            }
        }

        // Find description column
        for header in &headers {
            let lower = header.to_lowercase();
            if Some(header) != mapping.date.as_ref() {
                if desc_patterns.iter().any(|p| lower.contains(p)) {
                    mapping.description = Some(header.clone());
                    break;
                }
            }
        }

        // Fallback for description
        if mapping.description.is_none() {
            let fallback = ["name", "type", "ref", "reference", "category"];
            for header in &headers {
                let lower = header.to_lowercase();
                if Some(header) != mapping.date.as_ref() {
                    if fallback.iter().any(|p| lower.contains(p)) {
                        mapping.description = Some(header.clone());
                        break;
                    }
                }
            }
        }

        ServiceResult::ok(mapping)
    }

    /// Get the headers from a CSV file
    pub fn get_headers(file_path: &str) -> ServiceResult<Vec<String>> {
        let path = Path::new(file_path);
        if !path.exists() {
            return ServiceResult::fail(format!("File not found: {}", file_path));
        }

        let file = match File::open(path) {
            Ok(f) => f,
            Err(e) => return ServiceResult::fail(format!("Failed to open file: {}", e)),
        };

        let mut reader = csv::Reader::from_reader(BufReader::new(file));
        match reader.headers() {
            Ok(h) => ServiceResult::ok(h.iter().map(|s| s.to_string()).collect()),
            Err(e) => ServiceResult::fail(format!("Failed to read headers: {}", e)),
        }
    }

    /// Parse transactions from CSV file
    pub fn get_transactions(
        file_path: &str,
        mapping: &ColumnMapping,
        flip_signs: bool,
        debit_negative: bool,
    ) -> ServiceResult<Vec<Transaction>> {
        let path = Path::new(file_path);
        if !path.exists() {
            return ServiceResult::fail(format!("File not found: {}", file_path));
        }

        let file = match File::open(path) {
            Ok(f) => f,
            Err(e) => return ServiceResult::fail(format!("Failed to open file: {}", e)),
        };

        let mut reader = csv::Reader::from_reader(BufReader::new(file));

        // Get headers before iterating records
        let headers: Vec<String> = match reader.headers() {
            Ok(h) => h.iter().map(|s| s.to_string()).collect(),
            Err(e) => return ServiceResult::fail(format!("Failed to read headers: {}", e)),
        };

        let mut transactions = Vec::new();
        let now = Utc::now();

        for result in reader.records() {
            let record = match result {
                Ok(r) => r,
                Err(_) => continue, // Skip invalid rows
            };

            // Parse date
            let date_col = mapping.date.as_ref();
            let date_str = date_col.and_then(|col| {
                headers.iter().position(|c| c == col)
                    .and_then(|idx| record.get(idx))
            });

            let transaction_date = match date_str {
                Some(s) => match Self::parse_date(s) {
                    Some(d) => d,
                    None => continue,
                },
                None => continue,
            };

            // Parse amount
            let amount = if let Some(ref amt_col) = mapping.amount {
                let amt_str = headers.iter().position(|c| c == amt_col)
                    .and_then(|idx| record.get(idx));
                match amt_str.and_then(Self::parse_amount) {
                    Some(a) => a,
                    None => continue,
                }
            } else {
                // Handle debit/credit columns
                let debit_str = mapping.debit.as_ref().and_then(|col| {
                    headers.iter().position(|c| c == col)
                        .and_then(|idx| record.get(idx))
                });
                let credit_str = mapping.credit.as_ref().and_then(|col| {
                    headers.iter().position(|c| c == col)
                        .and_then(|idx| record.get(idx))
                });

                let debit_amt = debit_str.and_then(Self::parse_amount);
                let credit_amt = credit_str.and_then(Self::parse_amount);

                match (debit_amt, credit_amt) {
                    (Some(d), Some(c)) => {
                        if d.abs() > c.abs() { d } else { c }
                    }
                    (Some(mut d), None) => {
                        if debit_negative && d > Decimal::ZERO {
                            d = -d;
                        }
                        d
                    }
                    (None, Some(c)) => c,
                    (None, None) => continue,
                }
            };

            // Apply sign flip
            let final_amount = if flip_signs { -amount } else { amount };

            // Parse description
            let desc_str = mapping.description.as_ref().and_then(|col| {
                headers.iter().position(|c| c == col)
                    .and_then(|idx| record.get(idx))
            });
            let description = desc_str.map(|s| Self::clean_description(s));

            let transaction = Transaction {
                id: Uuid::new_v4(),
                account_id: Uuid::nil(), // Will be set by import service
                external_ids: HashMap::new(),
                amount: final_amount,
                description,
                transaction_date,
                posted_date: transaction_date,
                tags: Vec::new(),
                created_at: now,
                updated_at: now,
                deleted_at: None,
                parent_transaction_id: None,
            };

            transactions.push(transaction);
        }

        ServiceResult::ok(transactions)
    }

    /// Preview first N transactions
    pub fn preview_transactions(
        file_path: &str,
        mapping: &ColumnMapping,
        limit: usize,
        flip_signs: bool,
        debit_negative: bool,
    ) -> ServiceResult<Vec<Transaction>> {
        let result = Self::get_transactions(file_path, mapping, flip_signs, debit_negative);
        match result.data {
            Some(txs) => ServiceResult::ok(txs.into_iter().take(limit).collect()),
            None => result,
        }
    }

    fn parse_date(date_str: &str) -> Option<NaiveDate> {
        let s = date_str.trim();
        if s.is_empty() {
            return None;
        }

        let formats = [
            "%Y-%m-%d",  // 2024-10-01
            "%m/%d/%Y",  // 10/01/2024
            "%d/%m/%Y",  // 01/10/2024
            "%Y/%m/%d",  // 2024/10/01
            "%m-%d-%Y",  // 10-01-2024
            "%d-%m-%Y",  // 01-10-2024
        ];

        for fmt in &formats {
            if let Ok(date) = NaiveDate::parse_from_str(s, fmt) {
                return Some(date);
            }
        }

        None
    }

    fn parse_amount(amount_str: &str) -> Option<Decimal> {
        let s = amount_str.trim();
        if s.is_empty() {
            return None;
        }

        let mut cleaned = s.replace("$", "")
            .replace(",", "")
            .replace(" ", "");

        // Handle parentheses notation: (100.00) -> -100.00
        if cleaned.starts_with('(') && cleaned.ends_with(')') {
            cleaned = format!("-{}", &cleaned[1..cleaned.len()-1]);
        }

        Decimal::from_str(&cleaned).ok()
    }

    fn clean_description(description: &str) -> String {
        let mut cleaned = description.to_string();

        // Remove literal "null" strings
        let null_re = Regex::new(r"(?i)\bnull\b").unwrap();
        cleaned = null_re.replace_all(&cleaned, "").to_string();

        // Remove card number masks
        let card_re = Regex::new(r"(?i)x{10,}\d+").unwrap();
        cleaned = card_re.replace_all(&cleaned, "").to_string();

        // Clean up whitespace
        let ws_re = Regex::new(r"\s+").unwrap();
        cleaned = ws_re.replace_all(&cleaned, " ").trim().to_string();

        cleaned
    }

    /// Check if debits should be negated
    pub fn should_negate_debits(file_path: &str, debit_col: &str) -> ServiceResult<bool> {
        let path = Path::new(file_path);
        if !path.exists() {
            return ServiceResult::fail(format!("File not found: {}", file_path));
        }

        let file = match File::open(path) {
            Ok(f) => f,
            Err(e) => return ServiceResult::fail(format!("Failed to open file: {}", e)),
        };

        let mut reader = csv::Reader::from_reader(BufReader::new(file));
        let headers: Vec<String> = match reader.headers() {
            Ok(h) => h.iter().map(|s| s.to_string()).collect(),
            Err(e) => return ServiceResult::fail(format!("Failed to read headers: {}", e)),
        };

        let debit_idx = headers.iter().position(|h| h == debit_col);
        if debit_idx.is_none() {
            return ServiceResult::ok(false);
        }
        let idx = debit_idx.unwrap();

        let mut debit_values = Vec::new();
        for (i, result) in reader.records().enumerate() {
            if i >= 10 { break; } // Sample first 10 rows
            if let Ok(record) = result {
                if let Some(val) = record.get(idx) {
                    if let Some(amt) = Self::parse_amount(val) {
                        debit_values.push(amt);
                    }
                }
            }
        }

        if debit_values.len() >= 2 {
            let all_positive = debit_values.iter().all(|&amt| amt > Decimal::ZERO);
            return ServiceResult::ok(all_positive);
        }

        ServiceResult::ok(false)
    }
}
